File: easy/0872_leaf_similar_trees/solution.py
from typing import Optional

from utils.python.nodes import TreeNode


class Solution:
    def leafSimilar(
        self, root1: Optional[TreeNode], root2: Optional[TreeNode]
    ) -> bool:
        leafs1 = []
        leafs2 = []

        for root, leafs in ((root1, leafs1), (root2, leafs2)):
            stack = [root]

            while stack:
                node = stack.pop()

                if not node.left and not node.right:
                    leafs.append(node.val)
                    continue

                if node.left:
                    stack.append(node.left)
                if node.right:
                    stack.append(node.right)

        return leafs1 == leafs2

File: easy/0872_leaf_similar_trees/solution.cpp
#include "../../utils/cpp/nodes.hpp"
#include <vector>

class Solution {
public:
  bool leafSimilar(utils::TreeNode *root1, utils::TreeNode *root2) {
    std::vector<int> leafs1, leafs2;

    utils::TreeNode *node;

    for (auto &[root, leafs] :
         std::vector<std::pair<utils::TreeNode *, std::vector<int> *>>{
             {root1, &leafs1}, {root2, &leafs2}}) {
      std::vector<utils::TreeNode *> stack = {root};

      while (!stack.empty()) {
        node = stack.back();
        stack.pop_back();

        if (!node->left && !node->right) {
          leafs->push_back(node->val);
          continue;
        }

        if (node->left)
          stack.push_back(node->left);
        if (node->right)
          stack.push_back(node->right);
      }
    }

    return leafs1 == leafs2;
  }
};
